[quant][graphmode] Add InsertPackUnpack pass

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26939 [quant][graphmode] Integrate prepacked workaround in QuantFusion
* #26579 [quant][graphmode] Fold prepacked weight into module
* **#26959 [quant][graphmode] Add InsertPackUnpack pass**

Summary:
Add insert_pack_unpack pass for future transformations.
Only added a pattern for linear, will need to have a
pattern for conv2d as well

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636400](https://our.internmc.facebook.com/intern/diff/D17636400)

jerryzh168 committed:
[quant][graphmode] Add InsertPackUnpack pass

Summary:
Add insert_pack_unpack pass for future transformations.
Only added a pattern for linear, will need to have a
pattern for conv2d as well

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add InsertPackUnpack pass"


Summary:
Add insert_pack_unpack pass for future transformations.
Only added a pattern for linear, will need to have a
pattern for conv2d as well

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1176,6 +1176,10 @@ def get_forward(m):
 
+    @_tmp_donotuse_dont_inline_everything
+    def test_insert_pack_unpack(self):
+        pass
```
That's not a very good test :)

ZolotukhinM commented on torch/csrc/jit/passes/quantization.h:
```diff
@@ -118,6 +118,14 @@ TORCH_API void FoldConvBatchNorm2d(const script::Module& module);
 TORCH_API void FoldQuantizeCallIntoBuffer(script::Module& module, const std::string& method_name);
 
+/** \brief Insert pack and unpack function in all graphs
+ *   of module
```
This comment should probably be expanded with more details - otherwise it just repeats the function name, which is not helpful.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.h:
```diff
@@ -118,6 +118,14 @@ TORCH_API void FoldConvBatchNorm2d(const script::Module& module);
+
+/** \brief Insert pack and unpack function in graph
+ */
+TORCH_API void InsertPackUnpack(std::shared_ptr<Graph>& graph);
```
This doesn't need to be an external API.

jerryzh168 replied:
hmm, I want this for easier testing, since this is just a graph transformation

ZolotukhinM replied:
That's reasonable, just update the comment then.

ZolotukhinM commented on torch/csrc/jit/init.cpp:
```diff
@@ -180,9 +181,12 @@ void initJITBindings(PyObject* module) {
       .def(
           "_jit_pass_quant_fusion",
-          [](std::shared_ptr<Graph>& g) { return QuantFusion(g); })
+          [](std::shared_ptr<Graph> g) { return QuantFusion(g); })
```
What caused this change?

jerryzh168 replied:
I changed this for consistency with other passes, but I look again and saw that there are some APIs that passes `std::shared_ptr<Graph>` and some passes `std::shared_ptr<Graph>&`.

ZolotukhinM replied:
> I changed this for consistency with other passes, but I look again and saw that there are some APIs that passes `std::shared_ptr<Graph>` and some passes `std::shared_ptr<Graph>&`.

Majority of the passes uses the reference option, so let's use it here as well. Technically, it also saves one refcount increment.

jerryzh168 replied:
yeah, we should use `&` option

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -851,5 +851,41 @@ graph(%self, %scale, %zero_point, %dtype):
   rewriter.runOnGraph(graph, filter);
 }
+
+void InsertPackUnpack(std::shared_ptr<Graph> graph) {
```
This should be `std::shared_ptr<Graph>& graph` for consistency with other passes (and the declaration in the header file).

ZolotukhinM reviewed (changes requested):
Thanks for splitting this into a separate pass! Please find some comments inline.

jerryzh168 committed:
Update on "[quant][graphmode] Add InsertPackUnpack pass"


Summary:
Add insert_pack_unpack pass for future transformations.
Only added a pattern for linear, will need to have a
pattern for conv2d as well

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add InsertPackUnpack pass"


Summary:
Add insert_pack_unpack pass for future transformations.
Only added a pattern for linear, will need to have a
pattern for conv2d as well

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM reviewed (approved):
Looks good!

jerryzh168 committed:
Update on "[quant][graphmode] Add InsertPackUnpack pass"


Summary:
Add insert_pack_unpack pass for future transformations.
Only added a pattern for linear, will need to have a
pattern for conv2d as well

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636400](https://our.internmc.facebook.com/intern/diff/D17636400)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add InsertPackUnpack pass"


Summary:
Add insert_pack_unpack pass for future transformations.
Only added a pattern for linear, will need to have a
pattern for conv2d as well

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636400](https://our.internmc.facebook.com/intern/diff/D17636400)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add InsertPackUnpack pass"


Summary:
Add insert_pack_unpack pass for future transformations.
Only added a pattern for linear, will need to have a
pattern for conv2d as well

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636400](https://our.internmc.facebook.com/intern/diff/D17636400)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add InsertPackUnpack pass"


Summary:
Add insert_pack_unpack pass for future transformations.
Only added a pattern for linear, will need to have a
pattern for conv2d as well

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17636400](https://our.internmc.facebook.com/intern/diff/D17636400)

[ghstack-poisoned]
